src/adapter/src/coord/ddl.rs

```diff
@@ -86,6 +86,56 @@ impl Coordinator {
         self.catalog_transact_with(conn_id, ops, |_| Ok(())).await
     }
 
+    /// Executes a Catalog transaction with handling if the provided `Session` is in a SQL
+    /// transaction that is executing DDL.
+    pub(crate) async fn catalog_transact_with_ddl_transaction(
+        &mut self,
+        session: &mut Session,
+        ops: Vec<catalog::Op>,
+    ) -> Result<(), AdapterError> {
+        let conn_id = session.conn_id().clone();
+        let Some(Transaction {
+            ops:
+                TransactionOps::DDL {
+                    ops: txn_ops,
+                    state: txn_state,
+                    revision: txn_revision,
+                },
+            ..
+        }) = session.transaction_mut().inner_mut()
+        else {
+            return self.catalog_transact(Some(session), ops).await;
+        };
+
+        // Make sure our Catalog hasn't changed since openning the transaction.
+        if self.catalog().transient_revision() != *txn_revision {
+            return Err(AdapterError::DDLTransactionRace);
```

Comment:
Not covered, would be interesting if we can get here reproducibly, otherwise hopefully will be done by parallel-workload.

Reply:
I think the reason this isn't getting hit is because we bail [here](https://github.com/ParkMyCar/materialize/blob/catalog/minimal-transactional-ddl/src/adapter/src/session.rs#L1117-L1120), which is before we even run this code. The test that should exercises this scenario is [here](https://github.com/ParkMyCar/materialize/blob/catalog/minimal-transactional-ddl/test/sqllogictest/transactions.slt#L1032-L1049)